### ISSUE TITLE
Error in jped-cli

### ISSUE THREAD

**Issue by egoecho** [charite/jannovar#268]
When I run jped-cli (v0.16) there is an error. How to solve it? Thanks!

```
$ java -Xmx10G -jar /home/jxiong/programs/jannovar/jped-cli-0.16.jar \
    -d /home/jxiong/programs/jannovar/data/hg19_ensembl.ser \
    -m AUTOSOMAL_RECESSIVE \
    -p /home/jxiong/project/exome/family_analysis/ped_file_each_family/Family1.ped \
    -i /home/jxiong/project/exome/family_analysis/vcf_each_family/Family1.vcf \
    -o /home/jxiong/project/exome/family_analysis/output.vcf
verbosity: 1         
db path: /home/jxiong/programs/jannovar/data/hg19_ensembl.ser
ped path: /home/jxiong/project/exome/family_analysis/ped_file_each_family/Family1.ped
input path: /home/jxiong/project/exome/family_analysis/vcf_each_family/Family1.vcf
output path: /home/jxiong/project/exome/family_analysis/output.vcf
modeOfInheritance: AUTOSOMAL_RECESSIVE
geneWise: false      
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Exception in thread "main" java.lang.ClassCastException: de.charite.compbio.jannovar.
reference.GenomeInterval cannot be cast to java.lang.Comparable
        at com.google.common.collect.NaturalOrdering.compare(NaturalOrdering.java:26)
        at java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
        at java.util.TimSort.sort(TimSort.java:220)
        at java.util.Arrays.sort(Arrays.java:1512)
        at com.google.common.collect.ImmutableSortedSet.construct(ImmutableSortedSet.
java:376)            
        at com.google.common.collect.ImmutableSortedSet$Builder.build(ImmutableSorted
Set.java:515)        
        at com.google.common.collect.ImmutableSortedSet$SerializedForm.readResolve(Im
mutableSortedSet.java:773)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:
62)                  
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorIm
pl.java:43)          
        at java.lang.reflect.Method.invoke(Method.java:498)
        at java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1148)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1817)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1714)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1347)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1714)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1347)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2018)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1942)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1808)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:373)
        at de.charite.compbio.jannovar.data.JannovarDataSerializer.load(JannovarDataS
erializer.java:115)                                                                 
        at de.charite.compbio.jped.JPedApp.deserializeJannovarDB(JPedApp.java:86)
        at de.charite.compbio.jped.JPedApp.run(JPedApp.java:46)
        at de.charite.compbio.jped.JPedApp.main(JPedApp.java:93)
```

My java version is:
```
$java -version
openjdk version "1.8.0_91"
OpenJDK Runtime Environment (build 1.8.0_91-b14)
OpenJDK 64-Bit Server VM (build 25.91-b14, mixed mode)
```

**Comment by visze** [comment]
Thank you for the detailed description. There seems to be a problem while parsing the database file `hg19_ensembl.ser`.

Did you create the database with jannovar v0.16? It must always be the same version. Can you try to use other databases like refSeq or UCSC? Does the error also occur there?

You should also try to use the actual 0.17 version of jannovar. Please notice that the jped-cli functionality is moved to jannovar-cli in this version. Please let us know if the error also occur here.

**Comment by egoecho** [comment]
Thanks for your reply! I downloaded hg19_ensembl.ser using Jannovar-cli-0.17 but ran jped-cli-0.16. That may be the problem. May I ask how to use jped-cli function on jannovar-cli-0.17.jar? I tried below command but it is not work:

```
$java -Xmx4G -cp jannovar/jannovar-cli-0.17.jar de.charite.compbio.jannovar.filter
Error: Could not find or load main class de.charite.compbio.jannovar.filter
```

**Comment by visze** [comment]
OK. The serialized database is not backwards compatible.

Oh I see your problem! First of all it is not documented where to find it. And in addition the manual was not updated See [here](http://jannovar.readthedocs.io/en/master/jped.html)! 

But for you just run `java -jar jannovar-cli-0.17.jar annotate --help` first.

There you will see some options you need:

```
Usage: java -jar de.charite.compbio.jannovar.jar annotate [options] -d <database> -i [<IN.vcf>]+

...
  -d,--database <database.ser>       Jannovar database
...
  -i,--vcf-in <IN.vcf>               VCF file to annotate (.vcf/.gz)
...
     --pedigree-file <arg>           path to pedigree file
...
```
But this will just annotate the variants and not filter them! And I do not know if gene-wise filtering is always enabled or not (see recessive filter).

Sorry this is a big change and I was not aware of that! We have to think about that internally and will change and document it in the 0.18 release. Sorry about that.

Hope you can work with the "annotation" option. Or you have to create the database with jannovar v0.16.

@holtgrewe Let us talk about that after you vacation. I think that is a major issue!

**Comment by egoecho** [comment]
Thank you for your suggesion. I tried download database using jannovar-0.16 instead. However, a large mount of warning/error messages show up:

```
$ java -Xmx4G -jar jannovar-cli-0.16.jar download -d hg19/ensembl
... ( many lines as below )
INFO Transcript ENST00000505102 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000553057 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000551061 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000450232 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000443397 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000562234 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000605426 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000440875 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000444559 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000419938 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000409211 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000392861 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000431672 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000435831 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000427902 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000600684 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000520821 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000520028 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000419898 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000447581 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000469863 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000430882 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000413878 appears to be 3'/5' truncated. Ignoring.
INFO Transcript ENST00000418492 appears to be 3'/5' truncated. Ignoring.
ERROR Parsing FASTA...
[==================================================]   100%
INFO Found 170633 transcript models from Ensembl GFF resource, 137747 of which had se
quences                                                                             
INFO Serializing JannovarData to data/hg19_ensembl.ser
INFO Serialization took 30.24 sec.
```

Even though, file data/hg19_ensembl.ser was still created, and then I can use it for jped-cli-0.16.jar without error. Do you think is the hg19_ensembl.ser file OK and shall I just ignore these warning/error messages?

**Comment by visze** [comment]
Yes the the database file is fine. There are some strange transcripts in ensemble which are not "complete" that's why jannovar print out this error.
A short note: Jannovar db is not needed for your task, only if you enable the gene-wise option.

**Comment by egoecho** [comment]
I found another problem of jped-cli-0.16. In the case I used --gene-wise parameter (with hg19_ensembl.ser downloaded by jannovar-0.16), the output vcf file have only header but no any variant record. I've also tried added a "chr" prefix to the chromosome labels of my input vcf file but it is still the same. What's the problem it could be? Without --gene-wise, the jped-cli-0.16 has no such problem in either AUTOSOMAL_DOMINANT or AUTOSOMAL_RECESSIVE mode.

**Comment by visze** [comment]
What happens when you change the inheritance mode to AUTOSOMAL_RECESSIVE? Is the VCF file still empty? Gene-wise makes only sense on recessive inheritance using a gene as compound heterozygouse target.

**Comment by egoecho** [comment]
Yes, the output vcf is still empty.

> $ java -Xmx10G -jar /home/jxiong/programs/jannovar/jped-cli-0.16.jar \
> >     -d /home/jxiong/programs/jannovar/data-0.16/hg19_ensembl.ser \
> >     -m AUTOSOMAL_RECESSIVE --gene-wise \
> >     -p ../ped_file_each_family/Family12.ped \
> >     -i ../vcf_each_family_addchr/Family12.vcf \
> >     -o jpedFiltered_family12.vcf
> verbosity: 1
> db path: /home/jxiong/programs/jannovar/data-0.16/hg19_ensembl.ser
> ped path: ../ped_file_each_family/Family12.ped
> input path: ../vcf_each_family_addchr/Family12.vcf
> output path: jpedFiltered_family12.vcf
> modeOfInheritance: AUTOSOMAL_RECESSIVE
> geneWise: true
> SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
> SLF4J: Defaulting to no-operation (NOP) logger implementation
> SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
> Family used from PED file: 12
>     40
>     41
>     42
> Filtering and writing took 14.54 sec.
> $ tail -n5 jpedFiltered_family12.vcf
> ##contig=<ID=GL000225.1,length=211173>
> ##contig=<ID=GL000192.1,length=547496>
> ##reference=file:///home/jxiong/project/exome/data/genome/GRCh37-lite.fa
> ##source=SelectVariants
> #CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  40      41      42
> $

**Comment by visze** [comment]
can be if no variants match the inheritance pattern.

I can only check this if I have your ped and vcf files

**Comment by egoecho** [comment]
Sorry that it is not convenient to provide you the data (I have no permission). Is there any test data for jped-cli --gene-wise in github that I could have a try? 

In my input vcf file there are 655114 PASSed variants. Is it possible that none of them remain in --gene-wise mode? The ped file is:
```
$ cat Family12.ped
12      40      41      42      1       2
12      41      0       0       1       1
12      42      0       0       2       1
```

**Comment by visze** [comment]
Not likely. PED file seems fine!

Ok I tested it with short example. Same problem! So I am sorry. you have to switch to version 0.17 or later
```
##fileformat=VCFv4.1
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	40	41	42
1	874697	rs199789235	C	G	.	.	.	GT	0/1	0/1	0/0
1	874706	rs201028785	A	G	.	.	.	GT	0/1	0/0	1/0
1	874714	rs201596485	C	T	.	.	.	GT	1/1	0/1	0/1
```

**Comment by egoecho** [comment]
Thanks! So, how to use jped-cli funciton in jannovar 0.17? Could you give an example of the command line? In java -jar jannovar-cli-0.17.jar annotate --help, there is no parameter -m and --gene-wise. Or, is there jped-cli-0.17.jar can be downloaded on internet?

**Comment by visze** [comment]
We will describe it in the 0.18 release in more detail. Right now the documentation is missing. Sorry about that. 

The "new" inheritance CLI just annotates every position if it is dominant, recessive etc. There will be no filter. The resulting VCF have to be filtered afterwards yourself

**Comment by egoecho** [comment]
I tried below command in jannovar-cli-0.17 according to the --help. However, it always has an error:

```
$ java -Xmx2G -jar /home/jxiong/programs/jannovar/jannovar-cli-0.17.jar annotate -d /home/jxiong/programs/jannovar/data-0.17/hg19_ensembl.ser -i ../vcf_each_family/Family4.vcf --pedigree-file ../ped_file_each_family/Family1.ped --reference-fasta ../data/genome/GRCh37-lite.fa --show-all
ERROR: problem with parsing command line options: Could not parse the command line.

Use --help for obtaining usage instructions.
```

However, below command works fine:

`java -Xmx2G -jar /home/jxiong/programs/jannovar/jannovar-cli-0.17.jar annotate -d /home/jxiong/programs/jannovar/data-0.17/hg19_ensembl.ser -i ../vcf_each_family/Family4.vcf --show-all`

BTW, what is the --reference-fasta for? Shall I used it in my pedigree analysis?

**Comment by visze** [comment]
no. the reference-fasta is only needed for exac/1KG annotations

**Comment by egoecho** [comment]
But it still not work after I removed the --reference-fasta. What's wrong in my command?

```
$ java -Xmx2G -jar /home/jxiong/programs/jannovar/jannovar-cli-0.17.jar annotate -d /home/jxiong/programs/jannovar/data-0.17/hg19_ensembl.ser -i ../vcf_each_family/Family4.vcf --pedigree-file ../ped_file_each_family/Family1.ped --show-all
ERROR: problem with parsing command line options: Could not parse the command line.

Use --help for obtaining usage instructions.
```

**Comment by visze** [comment]
the command is `--showall` not `--show-all`and you do not need this because it is only used for the annotation with transcripts.

**Comment by egoecho** [comment]
Thanks so much! It finally works :) 

So, may I ask how the compound heterozygous mutation was labeled in the output vcf of jannovar-cli-0.17? Also, I found there are more than thousand variants annotated as "INHERITANCE=AD,AR", is that mean these mutations are compound heterozygous? (otherwise how could a mutation be both dominant and recessive?)

**Comment by visze** [comment]
So `AR` can be homozygous or comp heterozygous. What exactly you have to look at the genotypes (index `1/1` for homozygous).

`INHERITANCE=AD,AR` can also be possible especially if there are genotypes with an unknown allele (`.`). E.g the next example can be dominant (de-novo mutation) or homozygous:

```
index father mother
./1   0/.    0/.
```

**Comment by holtgrewe** [comment]
The annotation of compatible modes of inheritance is now documented in http://doc-openbio.readthedocs.io/projects/jannovar/en/develop/inheritance.html and will be included for the 0.18 documentation.
